extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheMethodValidator.java

```diff
@@ -0,0 +1,122 @@
+package io.quarkus.cache.deployment;
+
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_INVALIDATE;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_INVALIDATE_ALL;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_LOAD;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_STORE;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.NON_BLOCKING_RETURN_TYPES;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.MethodInfo;
+import org.jboss.jandex.Type;
+
+import io.quarkus.arc.processor.AnnotationStore;
+import io.quarkus.arc.processor.BeanInfo;
+import io.quarkus.cache.deployment.exception.MultipleCacheAnnotationsException;
+import io.quarkus.cache.deployment.exception.NoParameterException;
+import io.quarkus.cache.deployment.exception.NonBlockingReturnTypeException;
+import io.quarkus.cache.deployment.exception.VoidReturnTypeException;
+
+public class CacheMethodValidator {
+
+    public static void validateAnnotations(AnnotationStore annotationStore, BeanInfo bean, MethodInfo method,
+            List<Throwable> throwables) {
+
+        int methodAnnotationsCount = 0;
+        if (validateCacheLoadAnnotation(annotationStore, bean, method, throwables)) {
+            methodAnnotationsCount++;
+        }
+        if (validateCacheStoreAnnotation(annotationStore, bean, method, throwables)) {
+            methodAnnotationsCount++;
+        }
+        if (validateCacheInvalidateAnnotation(annotationStore, bean, method, throwables)) {
+            methodAnnotationsCount++;
+        }
+        if (validateCacheInvalidateAllAnnotation(annotationStore, bean, method, throwables)) {
+            methodAnnotationsCount++;
+        }
+        if (methodAnnotationsCount > 1) {
+            String error = "Multiple incompatible cache annotations found on the same method";
+            throwables.add(new MultipleCacheAnnotationsException(buildExceptionMessage(error, bean.getBeanClass(), method)));
+        }
+    }
+
+    private static boolean validateCacheLoadAnnotation(AnnotationStore annotationStore, BeanInfo bean, MethodInfo method,
+            List<Throwable> throwables) {
+        boolean found = false;
+        AnnotationInstance annotation = annotationStore.getAnnotation(method, CACHE_LOAD);
+        if (annotation != null) {
+            found = true;
+            checkVoidReturnType(CACHE_LOAD, bean, method).ifPresent(throwables::add);
+            checkNonBlockingReturnType(CACHE_LOAD, bean, method).ifPresent(throwables::add);
+            checkNoParameter(CACHE_LOAD, bean, method).ifPresent(throwables::add);
```

Comment:
I'm not sure how `@CacheInvalidate` would work if there's no key for `@CacheLoad` because we'd still need a key to invalidate the entry. We'd have to pass the `@CacheLoad`-annotated method name as the key parameter to the `@CacheInvalidate`-annotated method?

Reply:
That's a good question and the suggestion to use the method name was a bad idea (I thought about another option later and forgot to update the comment).

I think I would make it work this way:
- `@CacheLoad public Object load()`: creates an object with a specific chosen immutable key in the cache. Obviously, it makes more sense if the cache is named.
- `@CacheInvalidate public void invalidateCache()`: drops the object from this specific key.

Does it make sense or do I miss something obvious?

Reply:
Well, it's currently possible to have two (or more) `@CacheLoad`-annotated methods for the same cache (in the same class or in different classes) and these methods don't necessarily use the same type of key (the underlying cache accepts any `Object` as a key). I'm not sure the immutable key thing would work in such a situation.

Reply:
Well, basically, we would be able to support at most 1 method with no parameters for a given cache. Still better than requiring a parameter when you don't need one IMHO.

Reply:
BTW, Spring `@Cacheable` has a `key` parameter you can use to have a static key (well more than that but that's something you can do).

Reply:
With a one no-arg method per cache limit (which would be checked at build time), I think this could work. How would we avoid the collision risk between the immutable key and a user-chosen key?

Reply:
I think supporting `@CacheLoad` annotation without parameters could make sense. This could create an implicit key that could be used by the other `@CacheInvalidate`, as @gsmet mentions.

The only scenario that would be indeterministic would be if two key-less `@CacheLoad` (or any other `@Cache` annotation) existed for the same cache.

Reply:
> The only scenario that would be indeterministic would be if two key-less `@CacheLoad` (or any other `@Cache` annotation) existed for the same cache.

This could be detected and forbidden at build time, as well a documented properly in the javadoc and extension guide.

Reply:
I had some time to think about the indeterministic scenario thing while updating the code to allow key-less cached methods (it's almost done). It wouldn't actually be more indeterministic than a scenario with two `@CacheLoad`-annotated methods with the same key but different method bodies, which has been discussed in #3306 and is already possible with the current version of the code.

Reply:
I just pushed a new version which makes it possible to use cache annotations on no-args methods. At this point, the following combination in entirely possible and allowed in a bean:

```
public class Foo {
    @CacheLoad
    public String loadSomethingWithNoArg() {
        // method body here
    }
    @CacheLoad
    public String loadSomethingElseWithNoArg() {
        // method body here
    }
    @CacheLoad
    public String loadSomethingWithArg(Object key) {
        // method body here
    }
    @CacheLoad
    public String loadSomethingElseWithArg(Object key) {
        // method body here
    }
}

Tell me if this should be forbidden (if we want to keep things deterministic) and I'll add some checks at build time.
```

Reply:
@gwenneg Yeah, I think those are fine.